fairlearn/linear_model/_constrained_logistic.py

```diff
@@ -0,0 +1,813 @@
+        )
+        # concat with original data, drop the original sensitive_feature_ids
+        X = pd.concat([X, ohe_df], axis=1).drop(sensitive_feature_ids, axis=1)
+    else:  # Numpy array
+        transformed = enc.fit_transform(X[:, sensitive_feature_ids]).toarray()
+        # Delete the old column and append the transformed columns
+        X_without_sensitive = np.delete(X, sensitive_feature_ids, axis=1)
+        X = np.append(X_without_sensitive, transformed, axis=1)
+        # Need to return the new transformed sensitive feature ids since there are more columns now
+        renamed_sensitive_feature_ids = list(
+            range(X_without_sensitive.shape[1], X.shape[1])
+        )
+    categories = enc.categories_
+
+    return X, renamed_sensitive_feature_ids, categories
+
+
+class ConstrainedLogisticRegression(LogisticRegression):
+    r"""Constrained logistic regression.
+
+    This implementation closely follows the implementation in
+    :class:`sklearn.linear_model.LogisticRegression`, and is constrained
+    via the same way as in
+    `Zafar et al. (2017) <https://proceedings.mlr.press/v54/zafar17a.html>`_ [1]_.
+
+    The constraint is defined as the covariance between the sensitive
+    features :math:`\left\{\mathbf{z}_{i}\right\}_{i=1}^{N}`, and the
+    signed distance from the users` feature vectors to the decision
+    boundary :math:`\left\{d_{\boldsymbol{\theta}}\left(\mathbf{x}_{i}\right)\right\}_{i=1}^{N}`.
+    This is mathemetically described as follows:
+
+    .. math::
+      \operatorname{Cov}\left(\mathbf{z}, d_{\boldsymbol{\theta}}(\mathbf{x})\right)
+      =\mathbb{E}\left[(\mathbf{z}-\overline{\mathbf{z}})
+        d_{\boldsymbol{\theta}}(\mathbf{x})\right]
+        -\mathbb{E}[(\mathbf{z}-\overline{\mathbf{z}})]
+        \bar{d}_{\boldsymbol{\theta}}(\mathbf{x})\\
+      \approx \frac{1}{N} \sum_{i=1}^{N}\left(\mathbf{z}_{i}-\overline{\mathbf{z}}\right)
+      d_{\boldsymbol{\theta}}\left(\mathbf{x}_{i}\right)
+
+    The reasoning behind this derivation is further explained in [1]_.
+
+    Parameters
+    ----------
+    constraints : {'demographic_parity', 'none'}, default='demographic_parity'
+        The constraints used in the logistic regression. Currently,
+        only the constraint as in [1]_ is implemented via the
+        `'demographic_parity'` parameter. `'none'` will fall back to
+        the LogisticRegression as in sklearn.
+
+    penalty : {'l2', 'none'}, default='l2'
+        Specify the norm of the penalty:
+        - `'none'`: no penalty is added;
+        - `'l2'`: add a L2 penalty term and it is the default choice;
+
+    covariance_bound : float, default=0
+        The covariance bound that the constraint optimizes towards.
+        It can either be one of the following two:
+        - A single float that will be used for all sensitive features.
+        - A list of floats such that each sensitive feature has its own covariance bound.
```

Comment:
I've added the extra sentence now, and I will have a look at allowing a dictionary once I rewrite the code for supplying separate sensitive features.

Reply:
I am looking at this right now, and I ran into something that should be looked at instead I think. I now realize that we are asking for `covariance_bound` at the initialization of the `ConstrainedLogisticRegression`, but we are only asking for the `sensitive_features` when we are fitting the `ConstrainedLogisticRegression`. This peculiarity only now caught my eye, because it would be strange to allow for a dictionary with coviarance values at a time in which the `sensitive_features` are still unknown.

Would it be better to move `covariance_bound` to `.fit()`, such that the covariance values and the `sensitive_features` are provided at the same time? I think that would make the most sense for a user, but I am not 100% sure about it.

Reply:
I've had a similar train of thought! On a first glance it seemed a bit odd to me too, but I think the `covariance_bound` would be similar in spirit to things like `class_weight` in scikit-learn, which depends on class labels in a way that's similar as `covariance_bound` does for sensitive groups. This is also consistent with the current plans for `Thresholder`'s `group_to_upselect` (#997). 

As our scikit-learn API expert, @adrinjalali may have thoughts on this :)

Reply:
I'd need to read more, but the current design makes sense to me exactly for reasons that @hildeweerts mentions.

Reply:
Thanks for your explanation @hildeweerts! In that case, do you think it's still valuable to allow for a dictionary based on column names in case of a pandas.DataFrame? If so, I will look to do that.

Reply:
To me that would make sense!